Do not run jump-threading for GPUs

Flakebi opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149637)*
<!-- homu-ignore:end -->

GPU targets have convergent operations that must not be duplicated or
moved in or out of control-flow.
An example convergent operation is a barrier/syncthreads.

The only MIR pass affected by this is jump-threading, it can duplicate
calls. Disable jump-hreading for GPU targets to prevent generating
incorrect code.

This affects the amdgpu and nvptx targets.

Fixes rust-lang/rust#137086, see this issue for details.
Tracking issue: rust-lang/rust#135024

cc @RDambrosio016 @kjetilkjeka for nvptx
cc @ZuseZ4

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

bjorn3 commented on compiler/rustc_codegen_llvm/src/declare.rs:
```diff
@@ -68,6 +68,15 @@ pub(crate) fn declare_raw_fn<'ll, 'tcx>(
+            llvm::AttributePlace::Function,
+            &[llvm::AttributeKind::Convergent.create_attr(cx.llcx)],
+        );
+    }
```
Maybe put this in `attributes::apply_to_llfn` instead?

Flakebi replied:
Could you clarify what you mean?
I don’t see much to change here
1. I can’t put adding the convergent attribute anywhere lower-level as it requires knowing the target we are compiling for (we don’t want to pessimize non-gpu targets)
2. `apply_to_llfn` is used for function, parameter and return attributes, so the AttributePlace cannot be removed
3. Removing the `create_attr` isn’t trivial, `apply_to_llfn` would need to create a converted vector and there are some places where we directly put in an `Attribute` instead of an `AttributeKind`

Flakebi replied:
Yes, good point, I’ll do that.

Noratrieb commented on compiler/rustc_session/src/session.rs:
```diff
@@ -586,10 +586,15 @@ impl Session {
+            if self.target.has_convergent {
+                // Special care needs to be taken for convergent operations (i.e. not duplicating them).
+                // We do not want to handle these, so do not run any optimizations.
+                0
```
Someone shouldn't be able to override the MIR opt level to a higher one, enabling unsound optimizations.

Flakebi commented on compiler/rustc_target/src/spec/json.rs:
```diff
@@ -186,6 +186,7 @@ impl Target {
         forward_opt!(default_codegen_backend);
         forward!(trap_unreachable);
         forward!(requires_lto);
+        forward!(has_convergent);
```
Oh that makes sense, I’ll use is_like_gpu then. I don’t know of any other target needing convergent.

nnethercote commented:
cc @LegNeato @eddyb @FractalFir

nnethercote reviewed (commented):
Apologies, I reviewed this days ago but forgot to hit "submit review" :(

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149637/29e035e1728488dce5a03daa527ab1091d1cda47..f17636bc1bc10a2a552892241989f2de9fe006b4)*

RalfJung commented on compiler/rustc_session/src/session.rs:
```diff
@@ -586,10 +586,16 @@ impl Session {
+        if self.target.is_like_gpu {
+            // Special care needs to be taken for convergent operations (i.e. not duplicating them).
+            // We do not want to handle these, so do not run any optimizations.
+            0
```
That said, it seems wrong that force-inlining would be enabled for Miri,

Flakebi committed:
Do not run jump-threading for GPUs

GPU targets have convergent operations that must not be duplicated or
moved in or out of control-flow.
An example convergent operation is a barrier/syncthreads.

The only MIR pass affected by this is jump-threading, it can duplicate
calls. Disable jump-hreading for GPU targets to prevent generating
incorrect code.

This affects the amdgpu and nvptx targets.

folkertdev reviewed (commented):
The changes are only to comments, so I don't think a `stdarch` PR is needed.

One of the main reasons we want PRs over there is that we don't run `stdarch` CI in this repository at the moment.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149637/fd0c901b00ee1e08a250039cdb90258603497e20..074227f03631f54812922d5687b8aa63fcde2f8f)*

Flakebi commented on library/stdarch/crates/core_arch/src/amdgpu/mod.rs:
```diff
@@ -244,6 +244,13 @@ pub fn wavefrontsize() -> u32 {
+/// This operation is hence provided on a best-effort basis.
+/// Using it may result in incorrect code under some circumstances.
+///
+/// [convergent]: https://llvm.org/docs/ConvergentOperations.html
```
Hm, yes, is there a good way to combine with the existing comments?
Or do I need to do
```rust
#[doc = "description of function" + include_str!(…)]
```
or use concat?

```rust
#[doc = concat!("description of function\n", include_str!(…))]
```

Flakebi committed:
Document convergent limitation on intrinsics

Flakebi commented on library/stdarch/crates/core_arch/src/nvptx/mod.rs:
```diff
@@ -49,6 +49,8 @@ unsafe extern "C" {
 
 /// Synchronizes all threads in the block.
+///
+#[doc = include_str!("../amdgpu/intrinsic_is_convergent.md")]
```
Just to highlight this, amdgpu currently exposes most of these. As there is no common gpu folder, I decided to put the `.md` there and also include it here in nvptx. (Let me know if there is a better place!)

Flakebi reviewed (commented):
There are no new comments, so I hope everyone is (sufficiently) happy with this state :)

@rustbot ready

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149637/9df83179a4fd7ddee7532e72abb23c76b8426f9a..564a6822bf1a82977495b31d9f016b81de9a3ae0)*

nnethercote commented:
@bors r+
